pulley/src/regs.rs

```diff
@@ -1,181 +1,187 @@
-                    Self(index)
-                }
+        impl Reg for $reg_ty {
+            const RANGE: Range<u8> = $range;
 
-                /// Get this register's index.
-                #[inline]
-                pub fn index(&self) -> usize {
-                    usize::from(self.0)
-                }
+            unsafe fn new_unchecked(index: u8) -> Self {
+                core::mem::transmute(index)
             }
 
-            #[cfg(feature = "arbitrary")]
-            impl<'a> arbitrary::Arbitrary<'a> for $name {
-                fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
-                    let index = u.int_in_range(Self::RANGE.start..=Self::RANGE.end - 1)?;
-                    Ok(Self(index))
-                }
+            fn to_u8(self) -> u8 {
+                self as u8
             }
-        )*
-    }
+        }
+    };
 }
 
-define_registers! {
-    /// An `x` register: integers.
-    pub struct XReg = 0..37;
+/// An `s` register: integers.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[allow(non_camel_case_types)]
+pub enum SReg {
+    /// The special `sp` stack pointer register.
+    sp,
+
+    /// The special `lr` link register.
+    lr,
+
+    /// The special `fp` frame pointer register.
+    fp,
+
+    /// The special `spilltmp0` scratch register.
+    spilltmp0,
+
+    /// The special `spilltmp1` scratch register.
+    spilltmp1,
+}
 
-    /// An `f` register: floats.
-    pub struct FReg = 0..32;
+/// An `x` register: integers.
+#[repr(u8)]
+#[derive(Debug,Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[allow(non_camel_case_types, missing_docs)]
+#[rustfmt::skip]
+pub enum XReg {
+    x0,  x1,  x2,  x3,  x4,  x5,  x6,  x7,  x8,  x9,
+    x10, x11, x12, x13, x14, x15, x16, x17, x18, x19,
+    x20, x21, x22, x23, x24, x25, x26, x27, x28, x29,
+    x30, x31,
+}
 
-    /// A `v` register: vectors.
-    pub struct VReg = 0..32;
+/// An `f` register: floats.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[allow(non_camel_case_types, missing_docs)]
+#[rustfmt::skip]
+pub enum FReg {
+    f0,  f1,  f2,  f3,  f4,  f5,  f6,  f7,  f8,  f9,
+    f10, f11, f12, f13, f14, f15, f16, f17, f18, f19,
+    f20, f21, f22, f23, f24, f25, f26, f27, f28, f29,
+    f30, f31,
 }
 
+/// A `v` register: vectors.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[allow(non_camel_case_types, missing_docs)]
+#[rustfmt::skip]
+pub enum VReg {
+    v0,  v1,  v2,  v3,  v4,  v5,  v6,  v7,  v8,  v9,
+    v10, v11, v12, v13, v14, v15, v16, v17, v18, v19,
+    v20, v21, v22, v23, v24, v25, v26, v27, v28, v29,
+    v30, v31,
+}
```

Comment:
Why all these changes? Seems unrelated?

Reply:
Encodes the invariant that registers are always in the range `0..32`, so indexing into the registers array can never fail. The other option would be to use the `#[rustc_layout_scalar_valid_range_end(31)]` attribute, but that is nightly-only.

Reply:
Are you saying we get better codegen with `enum`s here? That's neat, I wouldn't have expected the LLVM IR that rustc emits to change.